simple/msg_netdir.c

```diff
@@ -0,0 +1,870 @@
+	case AF_INET6:
+		hints->addr_format = FI_SOCKADDR_IN6;
+		break;
+	}
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, opts.src_addr, opts.src_port,
+		FI_SOURCE, hints, &fi_pep);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		goto out;
+	}
+
+	minlen = min(strlen("netdir"), strlen(fi_pep->fabric_attr->prov_name)) + 1;
+	if (0 != (strncmp(fi_pep->fabric_attr->prov_name, "netdir", minlen))) {
+		FT_ERR("Unexpected provider - \"%s\", but expected provider is \"netdir\"",
+			fi_pep->fabric_attr->prov_name);
+		return -FI_EINVAL;
+	}
+
+	ret = fi_fabric(fi_pep->fabric_attr, &fabric, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_fabric", ret);
+		goto out;
+	}
+
+	ret = fi_eq_open(fabric, &eq_attr, &eq, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_eq_open", ret);
+		goto out;
+	}
+
+out:
+	freeaddrinfo(ai);
+	return ret;
+}
+
+static int run(void)
+{
+	char *node, *service;
+	uint64_t flags;
+	int ret;
+
+	ret = ft_read_addr_opts(&node, &service, hints, &flags, &opts);
+	if (ret)
+		return ret;
+
+	if (!opts.src_port)
+		opts.src_port = "9229";
+
+	if (!opts.src_addr) {
+		fprintf(stderr, "Source address (-s) is required for this test\n");
+		return -EXIT_FAILURE;
+	}
+
+	ret = setup_handle();
+	if (ret)
+		return ret;
+
+	if (!opts.dst_addr) {
+		ret = server_listen();
+		if (ret)
+			return ret;
+	}
+
+	ret = opts.dst_addr ? client_connect() : server_connect();
+	if (ret)
+		return ret;
+
+	ret = send_recv_greeting(ep);
+	if (ret)
+		return ret;
+
+	ret = send_recv_echo(ep);
+	if (ret)
+		return ret;
+
+	ret = check_cancelation(ep);
+	if (ret)
+		return ret;
+
+	ret = exchange_large_message(ep);
+	if (ret)
+		return ret;
+
+	ret = exchange_multiple_iovecs();
+	if (ret)
+		return ret;
+
+	ret = exchange_multiple_iovecs_via_rma();
+
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "h" CS_OPTS INFO_OPTS)) != -1) {
+		switch (op) {
+		default:
+			ft_parsecsopts(op, optarg, &opts);
+			ft_parseinfo(op, optarg, hints);
+			break;
+		case '?':
+		case 'h':
+			ft_usage(argv[0], "A simple MSG client-sever example based on "
+				"Network Direct.");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	hints->ep_attr->type = FI_EP_MSG;
+	hints->caps = FI_MSG | FI_RMA | FI_RECV | FI_SEND;
+	hints->mode = FI_CONTEXT;
+	hints->addr_format = FI_SOCKADDR;
+
+	ret = run();
+
+	if (mr != &no_mr)
+		FT_CLOSE_FID(mr);
+	FT_CLOSE_FID(ep);
+	FT_CLOSE_FID(txcq);
+	FT_CLOSE_FID(rxcq);
+	FT_CLOSE_FID(rxcntr);
+	FT_CLOSE_FID(txcntr);
+	FT_CLOSE_FID(domain);
+	FT_CLOSE_FID(eq);
+	FT_CLOSE_FID(fabric);
+
+	if (fi) {
+		fi_freeinfo(fi);
+		fi = NULL;
+	}
+
+	return ft_exit_code(ret);
+}
+#else /* _WIN32 */
+int main(int argc, char **argv)
+{
+	fprintf(stdout, "Network Direct provider works only under Windows OS \n");
+
+	return 0;
+}
+#endif /* !_WIN32 */
```

Comment:
fabtests should not have provider specific tests.  Any test should work over any provider that supports the necessary capabilities.  Please rename this test and submit separately if it needs to be included.  The name needs to reflect what the test is attempting to verify.  The test should do something not covered by other tests, which I'm not sure this does.

If you need to update the other tests to work over the ND provider (e.g. hook in WSAStartup somehow), please do that before adding a new test.